modules/lmdbbackend/lmdbbackend.cc

```diff
@@ -2620,17 +2620,100 @@ bool LMDBBackend::getTSIGKeys(std::vector<struct TSIGKey>& keys)
 
 string LMDBBackend::directBackendCmd(const string& query)
 {
-  if (query == "info") {
-    ostringstream ret;
+  ostringstream ret, usage;
 
+  usage << "info                               show some information about the database" << endl;
+  usage << "index check domains                check zone<>ID indexes" << endl;
+  usage << "index refresh domains <ID>         refresh index for zone with this ID" << endl;
+
+  vector<string> argv;
+  stringtok(argv, query);
+
+  if (argv.empty()) {
+    return usage.str();
+  }
+
+  string& cmd = argv[0];
+
+  if (cmd == "help") {
+    return usage.str();
+  }
+
+  if (cmd == "info") {
     ret << "shards: " << s_shards << endl;
     ret << "schemaversion: " << SCHEMAVERSION << endl;
 
     return ret.str();
   }
-  else {
-    return "unknown lmdbbackend command\n";
+
+  if (cmd == "index") {
+    if (argv.size() < 2) {
+      return "need an index subcommand\n";
+    }
+
+    string& subcmd = argv[1];
+
+    if (subcmd == "check") {
+      if (argv.size() < 3) {
+        return "need an index name\n";
+      }
+
+      if (argv[2] != "domains") {
+        return "can only check the domains index\n";
+      }
+
+      auto txn = d_tdomains->getROTransaction();
+
+      for (auto iter = txn.begin(); iter != txn.end(); ++iter) {
+        DomainInfo di = *iter;
+
+        auto id = iter.getID();
+
+        LMDBIDvec ids;
+        txn.get_multi<0>(di.zone, ids);
+
+        if (ids.size() != 1) {
+          ret << "ID->zone index has " << id << "->" << di.zone << ", ";
+
+          if (ids.empty()) {
+            ret << "zone->ID index has no entry for " << di.zone << endl;
+            ret << "  suggested remedy: index refresh domains " << id << endl;
+          }
+          else {
+            // ids.size() > 1
+            ret << "zone->ID index has multiple entries for " << di.zone << ": ";
+            for (auto id_ : ids) {
+              ret << id_ << " ";
+            }
+            ret << endl;
+          }
+        }
+      }
+      return ret.str();
+    }
+    if (subcmd == "refresh") {
+      // index refresh domains 12345
+      if (argv.size() < 4) {
+        return "usage: index refresh domains <ID>\n";
+      }
+
+      if (argv[2] != "domains") {
+        return "can only refresh in the domains index\n";
+      }
+
+      uint32_t id = atoi(argv[3].c_str());
```

Comment:
```suggestion
      auto id = checked_stoi<uint32_t>(argv[3]);
```

Reply:
I didn't do that because it says `stoi`, but you're saying it should just work?

Reply:
Yes, it should convert the string to the corresponding `uint32_t` value, or throw an exception if the string cannot be parsed as an unsigned integer value. Unless I'm missing something and you need a different behaviour?

Reply:
that sounds perfect. Thanks!